### PR TITLE
Feat: Trip Length Distributions

### LINKED ISSUE #22: [Feat]: Trip Length Distributions
Define functionality for producing static trip length distribution plots, e.g. with matplotlib. This should be able to use some of the existing matplotlib functionality and will just define a standard format for TLD plots.

The TLDs should be expected to be in the format defined in caf.toolkit.

A separate issue should be created for interactive TLD plots.

### PR DESCRIPTION
**Pull request by DanielJacksonTfN**
# Changes

Added function to plot a trip length distribution from a CostDistribution object. 

- Closes #22 

# Tasks

- [ ] Have unittests been added (testing individual functions and their edge cases)
- [ ] Has documentation (including user guide) been updated
- [ ] Have new dependencies been added (or changed) in relevant `requirements.txt`
- [ ] Code linting, tests and other workflows pass

### CODE REVIEW

File: tests/test_tld_plot.py [hunk 1]
```diff
@@ -0,0 +1,185 @@
+"""Testing for the plot_tld function in caf.viz.xy_plot."""
```
Comment (MattBuckley-TfN) [inline tests/test_tld_plot.py:1]: These unit tests look good, although there are a few conventions, and features, of pytest which worth bearing in mind for future. *I don't think these tests need changing here.*

- Convention: test files should be named based on the module they're testing e.g. `test_xy_plot`, and there should be one test file per module.
- Convention: tests should be grouped into classes based on the function within the module they're testing, e.g. all tests in should be invididual methods in a `TestPlotTld` class (instead of functions).
- Feature: instead of creating functions which are called at the start of tests to generate test data, or setup the test, pytest has [fixtures](https://docs.pytest.org/en/stable/explanation/fixtures.html) which are called automatically for any tests which have them as a parameter.
- Feature: (*not necessary here*) pytest has other decorators which can [mark tests](https://docs.pytest.org/en/stable/how-to/mark.html#mark), to either run different tests in different circumstances or run the same test with multiple different parameters ([`@pytest.mark.parametrize`](https://docs.pytest.org/en/stable/how-to/parametrize.html#parametrize-basics))

File: src/caf/viz/xy_plot.py [hunk 2]
```diff
@@ -24,6 +25,7 @@
     from collections.abc import Sequence
 
     import pandas as pd
+    from caf.toolkit.cost_utils import CostDistribution
```
Comment (MattBuckley-TfN) [inline src/caf/viz/xy_plot.py]: Prefer importing modules instead of classes, or functions, directly e.g. `from caf.toolkit import cost_utils` then `cost_utils.CostDistribution`. This is for two main reasons:
- Make it clearer where a function (or class) is defined when someone is reading the code
- Avoid adding additional things to documentation, sphinx generates documentation for any public functions / classes it finds in a module. Although I don't think this is a problem for imports behind TYPE_CHECKING

File: examples/run_tld_plot_example.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
+"""
+Trip Length Distribution Plot
+=====================================
+
+This example demonstrates how to create a Trip Length Distribution (TLD) plot using the
+`plot_tld` function from the `caf.viz.xy_plot` module.
```
Comment (MattBuckley-TfN) [inline examples/run_tld_plot_example.py]: These can be linked back to the functions using restructed text ```:ref:`link` ``` syntax
```suggestion
:func:`~caf.viz.xy_plot.plot_tld` function from the :mod:`caf.viz.xy_plot` module.
```

File: examples/run_tld_plot_example.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+"""
+
+# %%
+# The following are the required imports for this example
+import matplotlib.pyplot as plt
+import numpy as np
+import pandas as pd
+from caf.toolkit.cost_utils import CostDistribution
+
+from caf.viz.xy_plot import plot_tld
```
Comment (MattBuckley-TfN) [inline examples/run_tld_plot_example.py:16]: For some reason when this is ran using sphinx it doesn't use the TfN style, although it does work correctly when ran directly. A workaround is to add `plt.style.use("caf.viz.tfn")` at the top of the example (below imports), not sure exactly why this is required it's possibly something sphinx is doing behind the scenes.
```suggestion
from caf.viz.xy_plot import plot_tld

plt.style.use("caf.viz.tfn")
```

File: src/caf/viz/xy_plot.py [hunk 5]
```diff
@@ -376,3 +378,89 @@ def plot_xy(
...
+        total_trips = np.sum(cost_dist.trip_vals)
+
+        if total_trips > 0:
+            weighted_avg = (
+                np.sum(cost_dist.weighted_avg_vals * cost_dist.trip_vals) / total_trips
+            )
+
+            ax.axvline(
+                x=weighted_avg,
+                color="red",
```
Comment (MattBuckley-TfN) [inline src/caf/viz/xy_plot.py]: Instead of using specific colours try using "C1" to select a colour from the style
```suggestion
                color="C1",
```

File: src/caf/viz/xy_plot.py [hunk 6]
```diff
@@ -376,3 +378,89 @@ def plot_xy(
...
+                label=f"Weighted Avg: {weighted_avg:.2f}",
+            )
+            ax.legend()
+
+    # Plot the weighted average points for each bin as a scatter plot
+    if show_weighted_avg_points:
+        ax.scatter(
+            cost_dist.weighted_avg_vals,
+            cost_dist.trip_vals,
+            color="orange",
```
Comment (MattBuckley-TfN) [inline src/caf/viz/xy_plot.py]: Instead of using a specific colour, select one from the template with "C2"
```suggestion
            color="C2",
```

File: src/caf/viz/xy_plot.py [hunk 7]
```diff
@@ -376,3 +378,89 @@ def plot_xy(
...
+            label="Weighted Avg Points",
```
Comment (MattBuckley-TfN) [inline src/caf/viz/xy_plot.py]: Use the x label within this
```suggestion
            label=f"Weighted Avg. {xlabel}",
```

**Review by MattBuckley-TfN — CHANGES_REQUESTED** [review]
The code and example looks great, I've just got a few comments to make it use colours from the TfN style.

File: src/caf/viz/xy_plot.py (hunk 2 above)
Comment (DanielJacksonTfN) [reply src/caf/viz/xy_plot.py]: Noted, changed all instances of CostDistribution to cost_utils.CostDistribution and changed the import to import the module rather than the class.

File: tests/test_tld_plot.py (hunk 1 above)
Comment (DanielJacksonTfN) [reply tests/test_tld_plot.py:1]: Noted for future, cheers.

**Review by MattBuckley-TfN — APPROVED** [review]
LGTM
